flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AbstractAsyncRunnableStreamOperator.java

```diff
@@ -0,0 +1,123 @@
+        implements AsyncKeyOrderedProcessingOperator {
+
+    final KeySelector<?, ?> keySelector1;
+    final KeySelector<?, ?> keySelector2;
+    final ExecutorService asyncThreadPool;
+    final int asyncBufferSize;
+    final long asyncBufferTimeout;
+    final int inFlightRecordsLimit;
+
+    public AbstractAsyncRunnableStreamOperator(
+            KeySelector<?, ?> keySelector1,
+            KeySelector<?, ?> keySelector2,
+            ExecutorService asyncThreadPool,
+            int asyncBufferSize,
+            long asyncBufferTimeout,
+            int inFlightRecordsLimit) {
+        this.keySelector1 = keySelector1;
+        this.keySelector2 = keySelector2;
+        this.asyncThreadPool = asyncThreadPool;
+        this.asyncBufferSize = asyncBufferSize;
+        this.asyncBufferTimeout = asyncBufferTimeout;
+        this.inFlightRecordsLimit = inFlightRecordsLimit;
+    }
+
+    @Override
+    protected KeySelector getKeySelectorForAsyncKeyedContext(int index) {
+        switch (index) {
+            case 1:
+                return keySelector1;
+            case 2:
+                return keySelector2;
+            default:
+                throw new ArrayIndexOutOfBoundsException(
+                        "Try to get key selector for index " + index);
+        }
+    }
+
+    @Override
+    protected AsyncExecutionController createAsyncExecutionController() {
+        if (isAsyncKeyOrderedProcessingEnabled()) {
+            final StreamTask<?, ?> containingTask = checkNotNull(getContainingTask());
+            final MailboxExecutor mailboxExecutor =
+                    containingTask
+                            .getMailboxExecutorFactory()
+                            .createExecutor(getOperatorConfig().getChainIndex());
+            final int maxParallelism = environment.getTaskInfo().getMaxNumberOfParallelSubtasks();
+            return new SimpleAsyncExecutionController(
+                    mailboxExecutor,
+                    this::handleAsyncException,
+                    asyncThreadPool,
+                    getDeclarationManager(),
+                    getEpochParallelMode(),
+                    maxParallelism,
+                    asyncBufferSize,
+                    asyncBufferTimeout,
+                    inFlightRecordsLimit,
+                    null,
+                    getMetricGroup().addGroup("asyncProcessing"));
+        }
+        return null;
+    }
+
+    /**
+     * Define the parallel mode of the epoch manager. The default is {@link
+     * EpochManager.ParallelMode#SERIAL_BETWEEN_EPOCH}. Subclasses can override this method to
+     * change the parallel mode.
+     */
+    protected EpochManager.ParallelMode getEpochParallelMode() {
+        return EpochManager.ParallelMode.SERIAL_BETWEEN_EPOCH;
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <RET> AsyncFuture<RET> asyncProcess(CheckedSupplier<RET> runnable) {
+        return ((SimpleAsyncExecutionController) asyncExecutionController)
```

Comment:
check `isAsyncKeyOrderedProcessingEnabled` here?

Reply:
This is something like `asyncProcessWithKey`. The user call this method for async purpose, but if the isAsyncKeyOrderedProcessingEnabled == false ....  we dont know how to handle this